pkg/asset/installconfig/aws/metadata.go

```diff
@@ -106,25 +105,26 @@ func (m *Metadata) PrivateSubnets(ctx context.Context) (map[string]Subnet, error
 func (m *Metadata) PublicSubnets(ctx context.Context) (map[string]Subnet, error) {
 	err := m.populateSubnets(ctx)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "retrieving Public Subnets")
 	}
 
 	return m.publicSubnets, nil
 }
 
 func (m *Metadata) populateSubnets(ctx context.Context) error {
-	if len(m.publicSubnets) > 0 || len(m.privateSubnets) > 0 {
+	// If Subnets are already populated, no need to redo
+	if len(m.publicSubnets) > 0 || len(m.privateSubnets) > 0 || len(m.edgeSubnets) > 0 {
 		return nil
 	}
 
 	if len(m.Subnets) == 0 {
 		return errors.New("no subnets configured")
 	}
 
-	m.mutexSubnets.Lock()
-	defer m.mutexSubnets.Unlock()
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
```

Comment:
I'd rather have this function be dumber and just populate the subnets. Then we can drop the check if the subnets are populated and the mutex locking and move that to the calling functions. For example:
```golang
func (m *Metadata) PublicSubnets(ctx context.Context) (map[string]Subnet, error) {
    m.mutex.Lock()
    defer m.mutex.Unlock()

    if len(m.publicSubnets) == 0 {
	    err := m.populateSubnets(ctx)
	    if err != nil {
		    return nil, errors.Wrap(err, "retrieving Public Subnets")
        }
	}

	return m.publicSubnets, nil
}
```
With that change, all exported functions (InstanceTypes, VPC, *Subnets, AvailabilityZones, Session) will behave and look consistently. I think that's more desirable than a bit of code "duplication".

Reply:
I had considered that. This would also need an accompanying change (not just with mutexes) within populateSubnets().